src/main/java/seedu/address/logic/parser/SearchCommandParser.java

```diff
@@ -0,0 +1,139 @@
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import javafx.util.Pair;
+import seedu.address.logic.commands.SearchCommand;
+import seedu.address.logic.parser.exceptions.ParseException;
+import seedu.address.model.person.ContactContainsAllKeywordsPredicate;
+import seedu.address.model.person.ContactContainsAnyKeywordsPredicate;
+import seedu.address.model.person.ContactContainsKeywordsPredicate;
+import seedu.address.model.tag.Tag;
+
+/**
+ * Parses input arguments and creates a new SearchCommand object
+ */
+public class SearchCommandParser implements Parser<SearchCommand> {
+
+    /**
+     * Parses the given {@code String} of arguments in the context of the SearchCommand
+     * and returns an SearchCommand object for execution.
+     * @throws ParseException if the user input does not conform the expected format
+     */
+    public SearchCommand parse(String args) throws ParseException {
+        requireNonNull(args);
+        String trimmedArgs = args.trim();
+        if (trimmedArgs.isEmpty()) {
+            throw new ParseException(
+                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, SearchCommand.MESSAGE_USAGE));
+        }
+
+        ArgumentMultimap argMultimap =
+                ArgumentTokenizer.tokenize(args, PREFIX_NAME, PREFIX_PHONE, PREFIX_EMAIL, PREFIX_ADDRESS, PREFIX_TAG);
+
+        String condition = argMultimap.getPreamble().toLowerCase();
+
+        switch (condition) {
+        case SearchCommand.AND_CONDITION:
+            return parseSearchWithAndCondition(argMultimap);
+        case SearchCommand.OR_CONDITION:
+            return parseSearchWithOrCondition(argMultimap);
+        case SearchCommand.EMPTY_CONDITION:
+            return parseSearchWithEmptyCondition(argMultimap);
+        default:
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, SearchCommand.MESSAGE_USAGE));
+        }
+    }
+
+    private SearchCommand parseSearchWithAndCondition(ArgumentMultimap argMultimap) throws ParseException {
+        Pair<List<String>, List<List<String>>> keywordsAndPrefixes = extractPrefixesAndKeywords(argMultimap);
+        List<String> prefixes = keywordsAndPrefixes.getKey();
+        List<List<String>> keywords = keywordsAndPrefixes.getValue();
+        return new SearchCommand(new ContactContainsAllKeywordsPredicate(prefixes, keywords));
+    }
+
+    private SearchCommand parseSearchWithOrCondition(ArgumentMultimap argMultimap) throws ParseException {
+        Pair<List<String>, List<List<String>>> keywordsAndPrefixes = extractPrefixesAndKeywords(argMultimap);
+        List<String> prefixes = keywordsAndPrefixes.getKey();
+        List<List<String>> keywords = keywordsAndPrefixes.getValue();
+        return new SearchCommand(new ContactContainsAnyKeywordsPredicate(prefixes, keywords));
+    }
```

Comment:
```suggestion
    private SearchCommand parseSearchWithCondition(ArgumentMultimap argMultimap, boolean isAndCondition) throws ParseException {
        Pair<List<String>, List<List<String>>> keywordsAndPrefixes = extractPrefixesAndKeywords(argMultimap);
        List<String> prefixes = keywordsAndPrefixes.getKey();
        List<List<String>> keywords = keywordsAndPrefixes.getValue();
        return isAndCondition ? new SearchCommand(new ContactContainsAllKeywordsPredicate(prefixes, keywords))
            : new SearchCommand(new ContactContainsAnyKeywordsPredicate(prefixes, keywords));
    }
```
I realised there is some code duplication here, maybe can include one more parameter `isAndCondition` to decide whether we want to run `ContactContainsAllyKeywordsPredicate` or `ContactContainsAnyKeywordsPredicate`.

This is just a suggestion, if you think that this makes it clearer I think it looks good too :relaxed:

Reply:
Yup, I agree with that!